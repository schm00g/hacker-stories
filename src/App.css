.container {
  height: 100vw; 
  padding: 20px;
  background: #83a4d4; /* fallback for old browsers */ 
  background: linear-gradient(to right, #b6fbff65, #91b5f05d);
  color: #171212;
}

.headline-primary {
  font-size: 48px; 
  font-weight: 300; 
  letter-spacing: 2px;
}

.item {
  display: flex; 
  align-items: center; 
  padding-bottom: 5px;
}

.item > span {
  padding: 0 5px;
  white-space: nowrap; 
  overflow: hidden; 
  white-space: nowrap; 
  text-overflow: ellipsis;
}
  
.item > span > a { 
  color: inherit;
}
  
.button {
  background: transparent; 
  border: 1px solid #171212; 
  padding: 5px;
  cursor: pointer;
  transition: all 0.1s ease-in; 
}

.button:hover { 
  background: #171212; 
  color: #ffffff;
}

.button_small { 
  padding: 5px;
}

.button_large { 
  padding: 5px;
  border-radius: 2px;
  margin: 1%;
}

.search-form {
  padding: 10px 0 20px 0; 
  display: flex; 
  align-items: baseline;
}

.label {
  padding-left: 5px;
  font-size: 24px;
}

.input {
  border: none;
  border-bottom: 1px solid #171212; 
  background-color: transparent;
  font-size: 24px; 
}