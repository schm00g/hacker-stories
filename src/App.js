import * as React from 'react';
import axios from 'axios';

import './App.css';
import { ReactComponent as Check } from './check.svg';

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

const useSemiPersistentState = (key, initialState) => {
  const isMounted = React.useRef(false);

  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );
  
  React.useEffect(() => {
    if(!isMounted.current){
      isMounted.current = true;
    } else {
      localStorage.setItem(key, value);
    }
  }, [value, key]);

  return [value, setValue];
}

const storiesReducer = (state, action) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter((story) =>
          action.payload.objectID !== story.objectID
        )
      };
    default: 
      throw new Error();
  }
};

const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState('search', 'React');

  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);
  
  const [stories, dispatchStories] = React.useReducer(
    storiesReducer,
    {data: [], isLoading: false, isError: false}
  );

  const handleFetchStories = React.useCallback(async () => {
    dispatchStories({type: 'STORIES_FETCH_INIT'});

    try {
      const result = await axios.get(url);
  
      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: result.data.hits
      });
    } catch {
      dispatchStories({type: 'STORIES_FETCH_FAILURE'})
    }
  }, [url]);
  
  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = React.useCallback((item) => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item
    });
  }, []);
    
  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value);
  }
  
  const handleSearchSubmit = (event) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  }

  return (
    <div className="container">
      <h1 className="headline-primary">Hacker Stories</h1>
      <SearchForm 
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />
      <p>
        Searching for <strong>{searchTerm}</strong>
      </p>
      {stories.isError && <p>Something went wrong...</p>}
      {stories.isLoading ? (
        <p>Loading...</p>
        ) : (
          <List list={stories.data} onRemoveItem={handleRemoveStory}/>
        )}
    </div> 
  )
};

const InputWithLabel = ({
    id, 
    value, 
    type='text', 
    onInputChange,
    isFocused,
    children
  }) => {
    const inputRef = React.useRef();

    React.useEffect(() => {
      if(isFocused && inputRef.current){
        inputRef.current.focus();
      }
    }, [isFocused]);
  return (
    <>
      <label htmlFor={id} className="label">
        {children}
      </label>
      &nbsp;
      <input
        ref={inputRef} 
        id={id}
        type={type}
        value={value}
        onChange={onInputChange}
        className="input"
      />
    </>
  );
  }

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit
}) => (
  <form onSubmit={onSearchSubmit}>
    <InputWithLabel
      id="search"
      value={searchTerm}
      onInputChange={onSearchInput}
      isFocused
    >
        <strong>Search</strong>
    </InputWithLabel>
    <button 
      type="submit" 
      disabled={!searchTerm}
      className="button button_large"
    > 
      <Check height="18px" width="18px"/>
    </button>
  </form>
);

const List = React.memo(({list, onRemoveItem}) => (
    <div>
    <ul>
      {list.map((item) => (
        <Item 
        key={item.objectID} 
        item={item} 
        onRemoveItem={onRemoveItem}/>
      ))}
    </ul>
  </div>
  )
);

const Item = ({item, onRemoveItem}) => {
  return (
    <li className="item">
      <span style={{width: '50%'}}>
        <a target="_blank" href={item.url}>{item.title}</a>
      </span>
      <span style={{width: '15%'}}>@{item.author}</span>
      <span style={{width: '15%'}}>{item.num_comments} comments</span>
      <span style={{width: '10%'}}>{item.points} points</span>
      <span style={{width: '10%'}}>
        <button type="button" onClick={() => onRemoveItem(item)}>
          Remove
        </button>
      </span>
    </li>
  );
}

export default App;

// page 164
