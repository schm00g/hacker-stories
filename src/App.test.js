import * as React from 'react';
import App, {storiesReducer, SearchForm, InputWithLabel, List, Item} from './App.tsx';
import {
  render,
  screen,
  fireEvent,
  act
} from '@testing-library/react';

const storyOne = {
  title: 'Redux',
  url: 'https://redux.js.org/',
  author: 'Dan Abramov',
  num_comments: 2,
  points: 5,
  objectID: 1
}

const storyTwo = {
  title: 'React',
  url: 'https://reactjs.org/',
  author: 'Jorge Walke',
  num_comments: 3,
  points: 3.9,
  objectID: 2
}

const stories = [storyOne, storyTwo];

describe('something truthy and falsy', () => {
  test('true to be true', () => {
    expect(true).toBe(true);
  });

  test('expect false to be false', () => {
    expect(false).toBe(false);
  });
});


describe('storiesReducer', () => {
  test('removes a story from all stories', () => {
    const action = {type: 'REMOVE_STORY', payload: storyOne};

    const state = {data: stories, isLoading: false, isError: false};

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [storyTwo], 
      isLoading: false, 
      isError: false
    };

    expect(newState).toStrictEqual(expectedState);
  });

  test('initializes story fetch', () => {
    const action = {type: 'STORIES_FETCH_INIT'};

    const state = {data: [], isLoading: false, isError: false};

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [], 
      isLoading: true, 
      isError: false
    };

    expect(newState).toStrictEqual(expectedState);
  });

  test('stories fetch success', () => {
    const action = {type: 'STORIES_FETCH_SUCCESS', payload: stories};

    const state = {data: [], isLoading: true, isError: false};

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: stories, 
      isLoading: false, 
      isError: false
    };

    expect(newState).toStrictEqual(expectedState);
  });

  test('stories fetch failure', () => {
    const action = {type: 'STORIES_FETCH_FAILURE'};

    const state = {data: [], isLoading: true, isError: false};

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [], 
      isLoading: false, 
      isError: true
    };

    expect(newState).toStrictEqual(expectedState);
  })
});

describe('Item', () => {
  test('renders all properties', () => {
    render(<Item item={storyOne}/>);

    screen.debug();

    expect(screen.getByText('Dan Abramov')).toBeInTheDocument();
    expect(screen.getByText('Redux')).toHaveAttribute(
      'href',
      'https://redux.js.org/'
    );
  });

  test('renders a clickable remove button', () => {
    render(<Item item={storyTwo}/>);
    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  test('clicking the dismiss button calls the callback handler', () => {
    const handRemoveItem = jest.fn();

    render(<Item item={storyOne} onRemoveItem={handRemoveItem}/>);

    fireEvent.click(screen.getByRole('button'));

    expect(handRemoveItem).toHaveBeenCalledTimes(1);
  })
});

describe('Search Form', () => {
  const searchFormProps = {
    searchTerm: 'React',
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn()
  }

  test('renders the input field with its values', () => {
    render(<SearchForm {...searchFormProps}/>);

    // screen.debug();

    expect(screen.getByDisplayValue('React')).toBeInTheDocument();

    expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
  });

  test('calls onSearchInput on input field change', () => {
    render(<SearchForm {...searchFormProps}/>);

    fireEvent.change(screen.getByDisplayValue('React'), {
      target: {value: 'Redux'}
    });

    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
  });
})