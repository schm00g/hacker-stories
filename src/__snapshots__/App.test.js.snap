// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Form renders snapshot 1`] = `
<form>
  <label
    class="label"
    for="search"
  >
    <strong>
      Search
    </strong>
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button button_large"
    type="submit"
  >
     Submit
    <svg
      height="18px"
      width="18px"
    >
      check.svg
    </svg>
  </button>
</form>
`;
